Add explanations for why `canceled` has a return type of `F[Unit]` (#3107)
typelevel/cats-effect: docs/typeclasses/monadcancel.md

Self-cancelation is somewhat similar to raising an error with `raiseError` in that it will short-circuit evaluation and begin "popping" back up the stack until it hits a handler. Just as `raiseError` can be observed using the `onError` method, `canceled` can be observed using `onCancel`.

The primary differences between self-cancelation and `raiseError` are two-fold. First, `uncancelable` suppresses `canceled` within its body unless `poll`ed. For this reason, `canceled` has a return type of `F[Unit]` and not `F[Nothing]`:

```scala
IO.uncancelable { poll =>
  val cancelation: IO[Unit] = IO.canceled
  cancelation.flatMap { x =>

>     IO.println(s"This will print, meaning $x is not a Nothing")

Reviewer: ```suggestion
    IO.println(s"This will print, meaning $x cannot be Nothing")
```
